src/github.com/couchbase/sync_gateway/base/sharded_sequence_clock.go

```diff
@@ -255,13 +204,270 @@ func (s *ShardedClock) initPartition(partitionNo uint16) error {
+
+// ShardedClockPartition manages storage for one clock partition, where a clock partition is a set of
+// {vb, seq} values for a subset of vbuckets.
+// Modifying clock values and metadata is done directly to the []byte storage, to avoid marshal/unmarshal overhead.
+// SeqSize defines how many bytes are used to store each clock value.  It is initialized at 2 bytes/value (uint16 capacity), but gets increased
+// via the resize() operation when a call to SetSequence would exceed the current capacity.
+// Structure of []byte:
+//    index : 2 bytes.  Partition Index, as uint16.
+//    seqSize: 1 byte.  Sequence Size.  Supports values 1-4, where max sequence for that size is defined in kClockMaxSequences[size]
+//    vbucket sequences: 2-8 bytes per sequence (depending on seqSize)
+type ShardedClockPartition struct {
+	Key            string            // Clock partition document key
+	value          []byte            // Clock partition values
+	cas            uint64            // cas value of partition doc
+	vbucketOffsets map[uint16]uint16 // Offset for vbuckets
+	dirty          bool              // Whether values have been updated since last save/load
+}
+
+func NewShardedClockPartition(baseKey string, index uint16, vbuckets []uint16) *ShardedClockPartition {
+
+	initialSeqSize := 1
+
+	p := &ShardedClockPartition{
+		Key:   fmt.Sprintf(kClockPartitionKeyFormat, baseKey, index),
+		value: make([]byte, kShardedClockMetaSize+len(vbuckets)*initialSeqSize*2),
+	}
+	p.SetSeqSize(uint8(initialSeqSize))
+	p.SetIndex(index)
+	p.Init(vbuckets)
+	return p
+}
+
+func NewShardedClockPartitionForBytes(key string, bytes []byte, partitions *IndexPartitions) *ShardedClockPartition {
+	p := &ShardedClockPartition{
+		Key:   key,
+		value: bytes,
+	}
+	// load the index from the incoming bytes, and use it to get the vb set for this partition
+	partitionIndex := p.GetIndex()
+	vbuckets := partitions.PartitionDefs[partitionIndex].VbNos
+	p.Init(vbuckets)
+	return p
+}
+
+// Initializes vbucketOffsets
+func (p *ShardedClockPartition) Init(vbuckets []uint16) {
+
+	p.vbucketOffsets = make(map[uint16]uint16, len(vbuckets)-1)
+	for index, vbNo := range vbuckets {
+		p.vbucketOffsets[vbNo] = uint16(kShardedClockMetaSize) + 2*uint16(p.GetSeqSize())*uint16(index) // metadata + (2*size) bytes per vbucket
+	}
+}
+
+func (p *ShardedClockPartition) Marshal() ([]byte, error) {
+	return p.value, nil
+}
+
+func (p *ShardedClockPartition) Unmarshal(value []byte) error {
+	p.value = value
+	return nil
+}
+
+// Sets sequence.  Uses big endian byte ordering.
+func (p *ShardedClockPartition) SetSequence(vb uint16, seq uint64) {
+
+	if seq > kClockMaxSequences[p.GetSeqSize()] {
+		p.resize(seq)
+	}
+
+	p.setSequenceForOffset(p.vbucketOffsets[vb], p.GetSeqSize(), seq)
+}
+
+func (p *ShardedClockPartition) setSequenceForOffset(offset uint16, size uint8, seq uint64) {
+
+	switch size {
+	case 1:
```

Comment:
instead of raw numbers like "1", "2", can you define constants like seq_size_uint8, seq_size_uint16?  I think that will make the code slightly clearer.